src/repart/repart.c

```diff
@@ -5054,32 +5058,61 @@ static int partition_format_verity_sig(Context *context, Partition *p) {
         if (PARTITION_EXISTS(p))
                 return 0;
 
-        if (!context->private_key)
+        assert_se(hp = p->siblings[VERITY_HASH]);
+        assert(!hp->dropped);
+        assert_se(rp = p->siblings[VERITY_DATA]);
+        assert(!rp->dropped);
+
+        /* As per the https://uapi-group.org/specifications/specs/discoverable_partitions_specification/ the
+         * UUIDs of the data and verity partitions are respectively the first and second halves of the
+         * dm-verity roothash, so we can use them to match the signature to the right partition. */
+        if (!sd_id128_is_null(hp->current_uuid) && !sd_id128_is_null(rp->current_uuid)) {
+                uint8_t root_hash_key[sizeof(sd_id128_t) * 2];
+                memcpy(root_hash_key, rp->current_uuid.bytes, sizeof(sd_id128_t));
+                memcpy(root_hash_key + sizeof(sd_id128_t), hp->current_uuid.bytes, sizeof(sd_id128_t));
+                VeritySettings key = {
+                        .root_hash = &root_hash_key,
+                        .root_hash_size = sizeof(root_hash_key),
+                };
+                verity_settings = set_get(arg_verity_settings, &key);
+        }
```

Comment:
doesn't matter, but maybe split this out into its own helper, `VeritySettings* lookup_verity_settings_by_uuid_pair(sd_id128_t hash_uuid, sd_id128_t data_uuid)`

or so?